Fix bug in non-destructive `with type` with type constraint

rtjoa opened the pull request:
Non-destructive `with type` constraints can fail when a type constraint is present (when this should only happen for destructive substitution):
```ocaml
module type S = sig
  type ('a, 'b) t constraint 'a = 'l * 'r
end

type 'a t2 constraint 'a = 'l * 'r

module type S2 = S with type ('a, _) t = 'a t2
```
```
7 | module type S2 = S with type ('a, _) t = 'a t2
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Destructive substitutions are not supported for constrained
       types (other than when replacing a type constructor with
       a type constructor with the same arguments).
```

This appears to have been introduced in https://github.com/ocaml/ocaml/pull/13911.

cc @clementblaudeau @Octachron

rtjoa committed:
Add regression test

rtjoa committed:
Fix non-destructive `with type` with type constraints

clementblaudeau commented:
Good catch! The refactor of #13911 caused the error to be fired when building the replace function even when the destructive flag was set to `false`. I think more idiomatic solution to this issue would be to move the error outside of the `replace` function, just after the result of `merge`, something like:

```ocaml
    (* Merging *)
    let path, paths, tdecl, sg = merge ~patch ~destructive env sg loc lid in
    (* Preventing destructive substitution of constrained types, as the
       constraint would be lost *)
    let params = tdecl.typ_type.type_params in
    if destructive
       && params_are_constrained params
       && type_decl_is_alias sdecl = None then
      raise(Error(loc, env, With_cannot_remove_constrained_type));
    (* Post processing *)
    ...
```

Also, this issue demonstrates that the testing was insufficient (as #13911 passed all tests). It would be good to add a proper set of tests in `testsuite/tests/typing-modules/merge_constraint.ml`, that make sure all the conditions of the `if` are tested (including negative tests), maybe in the style of [this PR](https://github.com/ocaml/ocaml/pull/14100/files#diff-16475d6e2475e142459277f67dd1087fc001f8d9cf384c35c57c3eb49d67b6dc) ?

Octachron commented on typing/typemod.ml:
```diff
@@ -716,22 +716,25 @@ module Merge = struct
+            fun s path ->
+              Subst.Unsafe.add_type_function path ~params ~body s
+      else
+        fun s _ -> s
```
This part and the existence of the bug suggests an infelicity in the type of `post_process`. Since we don't use the replace argument if `destructive` is `false`, it might be better to replace the `destructive` argument by a `replace` argument which is `None` when `destructive` was `false` and `Some replace` otherwise. But it might be better to leave this for a subsequent PR.

rtjoa replied:
Good point. Leaving for a subsequent PR, to avoid conflicting with #14100.

Octachron commented:
The current position feels readable enough to me: we are raising an error when trying to build a `replace` function and failing to do so. And overall, we don't want to do any side-effects in a function that will be discarded.

Octachron reviewed (approved):
The fix is correct, and can be merged as it is, once the PR number is added to the changes entry for #13911 .

Nevertheless, @rtjoa if you wish to explore other refactorisation possibility, don't hesitate to say so.

rtjoa committed:
Add with type constraint testing

rtjoa committed:
Update changelog

rtjoa commented:
Thanks for the review. I updated the testsuite as suggested by @clementblaudeau and updated the change list. To keep this fix minimal, and because it could easily conflict with #14100, I'll leave further refactoring for future PRs.
